ultravox/data/datasets.py

```diff
@@ -1206,37 +605,328 @@ class Range(SizedIterableDataset):
     """Limits the number of samples from another dataset."""
 
     def __init__(
-        self,
-        dataset: data.IterableDataset,
-        num_samples: Optional[int] = None,
-        total_samples: Optional[int] = None,
+        self, dataset: SizedIterableDataset, num_samples: Optional[int] = None
     ) -> None:
         self._dataset = dataset
-        self._num_samples = num_samples
-
-        if isinstance(self._dataset, SizedIterableDataset):
-            self._estimated_length = len(self._dataset)
-        else:
-            if total_samples is None:
-                raise ValueError(
-                    "total_samples must be provided for non-SizedIterableDataset."
-                )
-            self._estimated_length = total_samples
-
-        if self._num_samples is not None and self._num_samples > self._estimated_length:
-            # Issuing a warning here instead of raising an error to accomodate for specific classes of VoiceDataset
-            # Once we migrate entirely to GenericVoiceDataset, we can raise an error here.
-            warnings.warn("num_samples is greater than total_samples.")
+        self._length = num_samples or len(dataset)
+        if self._length > len(dataset):
+            raise ValueError("num_samples exceeds dataset length.")
 
     def __iter__(self):
         for i, sample in enumerate(self._dataset):
-            if self._num_samples is not None and i >= self._num_samples:
+            if i >= self._length:
                 break
             yield sample
 
     def __len__(self):
-        return (
-            self._num_samples
-            if self._num_samples is not None
-            else self._estimated_length
-        )
+        return self._length
+
+
+CONTINUATION_USER_TEMPLATE = (
+    "Continue the following text using less than 50 words:\n\n<|audio|>"
+)
+CONTINUATION_ASSISTANT_TEMPLATE = "{{continuation}}"
+TRANSCRIPTION_USER_TEMPLATE = "Transcribe\n<|audio|>"
+
+BOOLQ_CONFIG = DatasetConfig(
+    path="fixie-ai/boolq-audio",
+    splits=[
+        DatasetSplitConfig(name="train", num_samples=10000),
+        DatasetSplitConfig(name="validation", num_samples=1000),
+    ],
+    user_template="{{passage}}\n\n{{'<|audio|>' if include_audio else question}}",
+    assistant_template="{{'True' if answer else 'False'}}",
+    transcript_template="{{question}}",
+)
```

Comment:
I'm wondering if we should put all of these dataset definitions in a `base_datasets.yaml` file. That way this file becomes way smaller too and only contains logic.

We just have to register them (e.g. at the end of this file) and we might need some variables, which IIRC yaml already has.

Reply:
I was fiddling with this and I think it can be factored out, but the imports started to get out of hand, will revisit in a future PR.